libs/server/Storage/Session/UnifiedStore/UnifiedStoreOps.cs

```diff
@@ -0,0 +1,318 @@
+        /// <param name="expireOption">Flags to use for the operation.</param>
+        /// <param name="unifiedContext">Basic context for the main store</param>
+        /// <param name="respCommand">The current RESP command</param>
+        /// <returns></returns>
+        public unsafe GarnetStatus EXPIRE<TUnifiedContext>(PinnedSpanByte key, long expiration, out bool timeoutSet, ExpireOption expireOption, ref TUnifiedContext unifiedContext, RespCommand respCommand)
+            where TUnifiedContext : ITsavoriteContext<UnifiedStoreInput, GarnetUnifiedStoreOutput, long, UnifiedSessionFunctions, StoreFunctions, StoreAllocator>
+        {
+            Span<byte> rmwOutput = stackalloc byte[OutputHeader.Size];
+            var unifiedOutput = new GarnetUnifiedStoreOutput(SpanByteAndMemory.FromPinnedSpan(rmwOutput));
+
+            // Convert to expiration time in ticks
+            var expirationTimeInTicks = respCommand switch
+            {
+                RespCommand.EXPIRE => DateTimeOffset.UtcNow.AddSeconds(expiration).UtcTicks,
+                RespCommand.PEXPIRE => DateTimeOffset.UtcNow.AddMilliseconds(expiration).UtcTicks,
+                RespCommand.EXPIREAT => ConvertUtils.UnixTimestampInSecondsToTicks(expiration),
+                _ => ConvertUtils.UnixTimestampInMillisecondsToTicks(expiration)
+            };
+
+            var expirationWithOption = new ExpirationWithOption(expirationTimeInTicks, expireOption);
+
+            var input = new UnifiedStoreInput(RespCommand.EXPIRE, arg1: expirationWithOption.Word);
+            var status = unifiedContext.RMW(key.ReadOnlySpan, ref input, ref unifiedOutput);
+
+            if (status.IsPending)
+                CompletePendingForUnifiedStoreSession(ref status, ref unifiedOutput, ref unifiedContext);
+
+            timeoutSet = status.Found && ((OutputHeader*)unifiedOutput.SpanByteAndMemory.SpanByte.ToPointer())->result1 == 1;
+
+            return status.Found ? GarnetStatus.OK : GarnetStatus.NOTFOUND;
+        }
+
+        /// <inheritdoc/>
+        public unsafe GarnetStatus RENAME(PinnedSpanByte oldKeySlice, PinnedSpanByte newKeySlice, bool withEtag)
+            => RENAME(oldKeySlice, newKeySlice, false, out _, withEtag);
+
+        /// <inheritdoc/>
+        public unsafe GarnetStatus RENAMENX(PinnedSpanByte oldKeySlice, PinnedSpanByte newKeySlice, out int result, bool withEtag)
+            => RENAME(oldKeySlice, newKeySlice, true, out result, withEtag);
+
+        private unsafe GarnetStatus RENAME(PinnedSpanByte oldKeySlice, PinnedSpanByte newKeySlice, bool isNX, out int result, bool withEtag)
+        {
+            result = -1;
+
+            // If same name check return early.
+            if (oldKeySlice.ReadOnlySpan.SequenceEqual(newKeySlice.ReadOnlySpan))
+            {
+                result = 1;
+                return GarnetStatus.OK;
+            }
+
+            // TODO verify the keys are in the same slot
```

Comment:
I think the appropriate place to validate is in `RespServerSession.CanServeSlot`, not once we're actually executing the command.